core/src/main/java/com/github/steanky/element/core/context/ElementContext.java

```diff
@@ -1,111 +1,531 @@
      *
      * @param <TElement> the type of the contextual element object
      * @return the element object
      */
     default <TElement> @NotNull TElement provide() {
-        return provide(null, DependencyProvider.EMPTY);
+        return provide(null, DependencyProvider.EMPTY, false);
     }
 
     /**
-     * Provides a contextual element object given a path key. If null, will attempt to provide the root node. This
-     * attempts to perform caching, if allowed to do so by the element object.
+     * Provides a collection of elements, given a valid path string pointing at a {@link ConfigList}, relative to this
+     * context's <i>root node</i>. This method catches {@link ElementException}s that are thrown when elements are
+     * provided, and relays them to the supplied consumer after the list has been iterated. This can allow certain
+     * elements to fail to load without preventing others from doing so.
      *
-     * @param path               the path key
+     * @param listPath           the path string pointing to the ConfigList
      * @param dependencyProvider the {@link DependencyProvider} used to provide dependencies
-     * @param <TElement>         the type of the element object
-     * @return the contextual element object
+     * @param cache              whether to prefer caching elements or not
+     * @param collectionSupplier the function used to create a {@link Collection} implementation based on a known size
+     * @param exceptionHandler   the consumer used to handle exceptions
+     * @param <TElement>         the type of element object
+     * @param <TCollection>      the type of collection
+     * @return a collection of provided element objects
      */
-    <TElement> @NotNull TElement provideAndCache(final @Nullable String path,
-            final @NotNull DependencyProvider dependencyProvider);
+    default @NotNull <TElement, TCollection extends Collection<TElement>> TCollection provideCollection(
+            final @NotNull String listPath, final @NotNull DependencyProvider dependencyProvider, final boolean cache,
+            final @NotNull IntFunction<? extends TCollection> collectionSupplier,
+            final @NotNull Consumer<? super ElementException> exceptionHandler) {
+        Objects.requireNonNull(listPath);
+        Objects.requireNonNull(dependencyProvider);
+        Objects.requireNonNull(collectionSupplier);
+        Objects.requireNonNull(exceptionHandler);
+
+        PathSplitter pathSplitter = pathSplitter();
+
+        Object[] ethylenePath = pathSplitter.splitPathKey(listPath);
+        String normalized = pathSplitter.normalize(listPath);
+
+        ConfigElement listElement = rootNode().getElement(ethylenePath);
+
+        if (listElement == null || !listElement.isList()) {
+            throw new ElementException("expected ConfigList at '" + normalized + "', found " + listElement);
+        }
+
+        ConfigList list = listElement.asList();
```

Comment:
why not getListOrThrow?

Reply:
getListOrThrow throws a ConfigProcessException which must be caught and subsequently rethrown, i want to use ElementException which is nicer for API usage as it is a runtime exception

Reply:
can't you wrap it